crates/lang_srv/src/server.rs

```diff
@@ -187,22 +296,264 @@ impl LanguageServer for RocLs {
+        let doc = params.text_document_position;
+        trace!("got completion request");
+        let res = panic_wrapper_async(|| async {
+            self.state
+                .registry
+                .completion_items(&doc.text_document.uri, doc.position)
+                .await
+        })
+        .await;
+        res
     }
 }
 
-fn panic_wrapper<T>(f: impl FnOnce() -> Option<T> + std::panic::UnwindSafe) -> Result<Option<T>> {
+async fn panic_wrapper_async<Fut, T>(
+    f: impl FnOnce() -> Fut + std::panic::UnwindSafe,
+) -> Result<Option<T>>
+where
+    Fut: Future<Output = Option<T>>,
+{
     match std::panic::catch_unwind(f) {
-        Ok(r) => Ok(r),
+        Ok(r) => Ok(r.await),
         Err(_) => Err(tower_lsp::jsonrpc::Error::internal_error()),
     }
 }
 
 #[tokio::main]
 async fn main() {
+    env_logger::Builder::from_env("ROCLS_LOG").init();
+
     let stdin = tokio::io::stdin();
     let stdout = tokio::io::stdout();
 
-    let (service, socket) = LspService::new(RocLs::new);
+    let (service, socket) = LspService::new(RocServer::new);
     Server::new(stdin, stdout, socket).serve(service).await;
 }
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        sync::{Once, OnceLock},
+        time::Duration,
+    };
+
+    use expect_test::expect;
+    use tokio::{join, spawn};
+
+    use super::*;
+
+    fn completion_resp_to_labels(resp: CompletionResponse) -> Vec<String> {
+        match resp {
+            CompletionResponse::Array(list) => list.into_iter(),
+            CompletionResponse::List(list) => list.items.into_iter(),
+        }
+        .map(|item| item.label)
+        .collect::<Vec<_>>()
+    }
+    ///Gets completion and returns only the label for each completion
+    async fn get_completion_labels(
+        reg: &Registry,
+        url: &Url,
+        position: Position,
+    ) -> Option<Vec<String>> {
+        reg.completion_items(url, position)
+            .await
+            .map(completion_resp_to_labels)
+    }
+
+    const DOC_LIT: &str = r#"
+app "fizz-buzz"
+    packages { pf: "https://github.com/roc-lang/basic-cli/releases/download/0.5.0/Cufzl36_SnJ4QbOoEmiJ5dIpUxBvdB3NEySvuH82Wio.tar.br" }
+    imports [pf.Stdout,pf.Task.{ Task, await },]
+    provides [main] to pf
+"#;
+    static INIT: Once = Once::new();
+    async fn test_setup(doc: String) -> (RocServerState, Url) {
+        INIT.call_once(|| {
+            env_logger::builder()
+                .is_test(true)
+                .filter_level(log::LevelFilter::Trace)
+                .init();
+        });
+        let url = Url::parse("file:/test.roc").unwrap();
+
+        let inner = RocServerState::new();
+        //setup the file
+        inner.change(&url, doc, 0).await.unwrap();
+        (inner, url)
+    }
+
+    #[tokio::test]
+    async fn test_completion_with_changes() {
+        let doc = DOC_LIT.to_string()
+            + r#"rec=\a,b->{one:{potato:\d->d,leak:59},two:b}
+rectest= 
+  value= rec 1 2
+  va"#;
```

Comment:
Can you update this to use the `indoc!` macro, to avoid the un-indentattion? Also, if it wouldn't be inconvinient, formatting this to be in the standard Roc format would be appreciated.